Monika After Story/game/zz_apikeys.rpy

```diff
@@ -0,0 +1,556 @@
+
+        RETURNS: True if the feature was added, False if not.
+        """
+        if store.mas_api_keys.feature_registered(feature):
+            # feature already registered
+            return False
+
+        store.mas_api_keys.register_feature(
+            feature,
+            (display_name, on_change)
+        )
+        return True
+
+
+init -980 python in mas_api_keys:
+    import json
+    import os
+    import pygame
+    import store
+    import store.mas_utils as mas_utils
+
+    import store.mas_api_keys_data as Data
+
+
+    def feature_registered(feature):
+        """
+        Checks if a feature is registered for API keys
+
+        IN:
+            feature - feature to check (string)
+
+        RETURNS: True if the feature is registered for API keys
+        """
+        return feature in Data.registered_api_keys
+
+
+    def register_feature(feature, data):
+        """
+        Registers a feature for API key usage
+
+        IN:
+            feature - name of the feature
+            data - data to associate with the feature
+        """
+        Data.registered_api_keys[feature] = data
+
+
+    def features_for_display():
+        """
+        Returns list of the features for display
+
+        RETURNS: list of the features as tuples, sorted for display:
+            [0] - the display name of the feature
+            [1] - the on_change function to run
+        """
+        feats = []
+        for feature in Data.registered_api_keys:
+            key = store.mas_getAPIKey(feature)
+
+            # reduce length to avoid going past screen edge
+            if len(key) > Data.MAX_KEY_SIZE_DISP:
+                key = key[:Data.MAX_KEY_SIZE_DISP - 3] + "..."
+
+            feats.append((
+                Data.registered_api_keys[feature][0],
+                feature,
+                key,
+            ))
+
+        return sorted(feats)
+
+
+    def has_features():
+        """
+        Do we have API-key based features?
+
+        RETURNS: True if we have api key based features
+        """
+        return len(Data.registered_api_keys) > 0
+
+
+    def clean_key(dirty_key):
+        """
+        Cleans an aPI key
+
+        IN:
+            dirty_key - key to clean
+
+        RETURNS: cleaned key
+        """
+        return dirty_key.replace("\n", "").replace("\r", "")
+
+
+    def _run_on_change(feature, api_key):
+        """
+        Runs on change for a feature with api key
+
+        IN:
+            feature - feature to run on change for
+            api_key - api key to run on change with
+
+        RETURNS: tuple of the following format:
+            [0] - True if valid key, False if not
+            [1] - error message to show
+        """
+        on_change = Data.registered_api_keys[feature][1]
+        if on_change is not None:
+            try:
+                rv = on_change(api_key)
+
+                if not isinstance(rv, tuple) or len(rv) < 2:
```

Comment:
`len(rv) < 2` is more pythonic since the returned value is only indexed at 0 and 1. If a larger tuple/list/indexable object is returned, it's not a big deal since it wont crash.

I'll add a type check for the values at 0 and 1, though.